### DISCUSSION TITLE
Offline OpenStreetMap in the cache

### DISCUSSION THREAD

**Discussion by MikhailDomanov** [merlos/iOS-Open-GPX-Tracker#242] [General]
Hello Mr Juan M. Merlos, Mr Vincent Neo and all the Team, 

This piece of software is just great!

I would like to have OpenStreetMap offline in the app's cache around my location. That is, withing the radius of 200 km. In the summer I travel a lot and change my plans depending on the weather, road conditions and many other factors. As I understand, the app allows to browse areas in advance and to save them into its cache. I cannot see an option to select an area and download all map tiles in all magnifications for a large area with a couple of taps in advance. 

I am not a software developer. I cannot quickly read this source code. Thus, I would like to ask if I could download a complete set of map tiles from OpenStreetMap on my PC with Windows, convert it into the cache format used by Open GPX Tracker (1.8.0). I will then put this cache into my (jailbroken) iPhone manually.

I am sure you know a couple of ways I could proceed. Any ideas will be good ideas!

Thank you

**Comment by merlos** [comment]
Hi Mikhail,

In order to do what you mention you need some basic level of software development or command line scripting. 

The process would be to:

1) Download the tiles in your pc saving each tile file should have as name the MD5 hash of the following string `urlTemplate-x-y-z` where `x` `y` `z` are the number of the tiles, and `urlTemplate`  depends on the map provider. You have the templates used in the app in this linkhttps://github.com/merlos/iOS-Open-GPX-Tracker/blob/4981e2d342e1c4807bb7e44d87f6401cc51be29c/OpenGpxTracker/GPXTileServer.swift#L53

For example the file with the URL:
https://tile.openstreetmap.org/6/32/22.png ---> shall be renamed to the md5 of "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png-32-22-6". For example in osx you can use md5 

```shell
md5 -s "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png-32-22-6" 
```
Which will give you the output
```
MD5 ("https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png-32-22-6") = 9da6c8c2c6dfed0c3f4369df6cd6a86a
```
Here 9da...86a is the name you have to assign to the file.

In this other link you have how the url and the key are calculated within the app
https://github.com/merlos/MapCache/blob/13283eaeec68951f330a1229e0bb4ae23e05c789/MapCache/Classes/MapCache.swift#LL53C1-L73C6

 2) Once you have the tiles. Copy the files with the md5 name to the temporary folder where the cache is created.  It is within the `Discache` folder within `Library/caches` in the app folder within your device (see [Apple doc](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-S)
 
That should be the process. I would check moving some tiles, and using the phone off-line to check if it works. Then, download the full tileset. 

Alternatively, in the `MapCache` repo, the [example app](https://github.com/merlos/MapCache/tree/13283eaeec68951f330a1229e0bb4ae23e05c789/Example) allows you to download a region in the map. If you can compile it, it may be an alternative. 
 
 Best, Juan.

**Reply by MikhailDomanov** [reply]
Juan, 

This is huge and EXACTLY what I was looking for. 
Estoy muy agradecido

Mikhail
